modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItTruncateSuffixAndRestartTest.java

```diff
@@ -0,0 +1,384 @@
+     */
+    @Test
+    void testRestartTwoNodes() throws Exception {
+        SimpleIgniteNode node0 = nodes.get(0);
+        SimpleIgniteNode node1 = nodes.get(1);
+        SimpleIgniteNode node2 = nodes.get(2);
+
+        assertThat(node1.getService().run(testWriteCommand("1")), willCompleteSuccessfully());
+        waitForValueReplication(nodes, "1");
+
+        node1.stopService();
+        node2.stopService();
+
+        node1.logStorage.truncateSuffix(1);
+
+        node2.startService();
+        node1.startService();
+
+        waitForValueReplication(nodes, "1");
+
+        assertThat(node1.getService().run(testWriteCommand("2")), willCompleteSuccessfully());
+        waitForValueReplication(nodes, "2");
+
+        assertEquals(node0.logStorage.getLastLogIndex(), node1.logStorage.getLastLogIndex());
+        assertEquals(node1.logStorage.getLastLogIndex(), node2.logStorage.getLastLogIndex());
+    }
+
+    private static String nodeName(int i) {
+        return "foo" + i;
+    }
+
+    private static int port(int i) {
+        return DEFAULT_PORT + i;
+    }
+
+    private static void waitForValueReplication(List<SimpleIgniteNode> nodes, String value) throws InterruptedException {
+        for (SimpleIgniteNode node : nodes) {
+            assertTrue(waitForCondition(() -> value.equals(node.raftGroupListener.lastValue), 2, 10_000));
+        }
+    }
+
+    /**
+     * Version of {@link OnHeapLogs} that recovers raft configuration on start, because it may not be empty at that time.
+     */
+    private static class ReusableOnHeapLogs extends OnHeapLogs {
+        @Override
+        public boolean init(LogStorageOptions opts) {
+            ConfigurationManager confManager = opts.getConfigurationManager();
+
+            // Log always starts with "1". There's also no truncation in this test.
+            assertNull(getEntry(0));
+
+            for (int i = 1;; i++) {
+                LogEntry entry = getEntry(i);
+
+                // Loop until there are no more log entries. That's the only way to iterate the on-heap log.
+                if (entry == null) {
+                    break;
+                }
+
+                if (entry.getType() == EnumOutter.EntryType.ENTRY_TYPE_CONFIGURATION) {
+                    ConfigurationEntry confEntry = new ConfigurationEntry();
+
+                    confEntry.setId(new LogId(entry.getId().getIndex(), entry.getId().getTerm()));
+                    confEntry.setConf(new Configuration(entry.getPeers(), entry.getLearners()));
+
+                    if (entry.getOldPeers() != null) {
+                        confEntry.setOldConf(new Configuration(entry.getOldPeers(), entry.getOldLearners()));
+                    }
+
+                    confManager.add(confEntry);
+                }
+            }
+
+            return super.init(opts);
+        }
+    }
+
+    private static class TestRaftGroupListener implements RaftGroupListener {
+        String lastValue;
```

Comment:
Shouldn't it be `volatile` ?

Reply:
I guess there is HB somewhere, but you're right, volatile should be explicit here